### PR TITLE
[expo-cli] check expo-updates is installed before a publish

### PR DESCRIPTION
**Pull request by jkhales**
# Why

As of sdk 44 we no longer include expo-updates.

# How

Prevent accidental publishes.

# Test Plan

Tested in demo repo.

### CODE REVIEW

File: packages/config/src/Config.ts [hunk 1]
```diff
@@ -140,9 +140,7 @@ export function getConfig(projectRoot: string, options: GetConfigOptions = {}):
       if (configWithDefaultValues.exp.hooks) {
         delete configWithDefaultValues.exp.hooks;
       }
-      if (configWithDefaultValues.exp.ios?.config) {
-        delete configWithDefaultValues.exp.ios.config;
-      }
+      delete configWithDefaultValues.exp.ios?.config;
```
Comment (brentvatne) [inline packages/config/src/Config.ts:143]: looks like unrelated change

File: packages/expo-cli/src/commands/publish/publishAsync.ts [hunk 2]
```diff
@@ -150,6 +154,25 @@ function assertValidReleaseChannel(releaseChannel?: string): void {
...
+  const packageJsonPath = resolveFrom.silent(projectDir, './package.json');
+  if (packageJsonPath) {
+    const expoPackageJson = JsonFile.read(packageJsonPath, { json5: true });
+    if (expoPackageJson.dependencies && (expoPackageJson.dependencies as any)['expo-updates']) {
+      return;
+    }
+  }
+
+  throw new Error(
+    `Expo CLI is not installed in this project. Please run "expo install expo-updates".`
```
Comment (brentvatne) [inline packages/expo-cli/src/commands/publish/publishAsync.ts:172]: - i believe we should warn and prompt instead of hard fail here to account for one of the possible strange edge cases, so we don't block anybody.
- i suspect you meant to write "expo-updates" and not "Expo CLI"
- i realize now that we don't have a docs page with info about installing expo-updates which accounts for both managed and bare projects. in managed apps people just need to run `expo install expo-updates` but in bare apps we may want to link them to [this docs page](https://docs.expo.dev/bare/installing-updates/)

File: packages/expo-cli/src/commands/publish/publishAsync.ts [hunk 3]
```diff
@@ -150,6 +154,25 @@ function assertValidReleaseChannel(releaseChannel?: string): void {
   }
 }
 
+function assertExpoUpdatesInstalled(projectDir: string, sdkVersion?: string): void {
+  // before sdk 44, expo-update was included in with the expo module
+  if (sdkVersion && semver.lt(sdkVersion, '44.0.0')) {
```
Comment (brentvatne) [inline packages/expo-cli/src/commands/publish/publishAsync.ts:159]: if someone is using expo-updates prior to sdk 44 in a bare project they would have react-native-unimodules instead of the expo package, and in that case their sdkVersion would be undefined. I'm not sure why someone would run `expo publish` in that context, though.

i'd prefer to skip any validation we're doing here in the context that the sdkVersion is undefined for now because we're specifically trying to address sdk 44+ defaulting to no expo-updates in this pr.

File: packages/expo-cli/src/commands/publish/publishAsync.ts [hunk 4]
```diff
@@ -150,6 +154,25 @@ function assertValidReleaseChannel(releaseChannel?: string): void {
...
+    return;
+  }
+
+  const packageJsonPath = resolveFrom.silent(projectDir, './package.json');
+  if (packageJsonPath) {
+    const expoPackageJson = JsonFile.read(packageJsonPath, { json5: true });
+    if (expoPackageJson.dependencies && (expoPackageJson.dependencies as any)['expo-updates']) {
+      return;
+    }
+  }
```
Comment (brentvatne) [inline packages/expo-cli/src/commands/publish/publishAsync.ts:169]: we have existing code for handling reading the package.json - in the `actionAsync` function above the `pkg` is returned from `getConfig`, for example. you could pass that pkg into this function instead of projectDir.

also, checking `dependencies` may not be a completely dependable way of determining whether the library is installed -- maybe it's accidentally installed as a dev dependency, in which case it would still work fine. there may be some edge cases in which the user is depending on expo-updates via a transitive dep (unlikely but with enough users is bound to happen) or they use something like nx for their monorepo, which i believe centralizes dependencies in a monorepo root package.json. see next comment for follow up on this point.

**Review by brentvatne — CHANGES_REQUESTED** [review]

File: packages/expo-cli/src/commands/publish/publishAsync.ts (hunk 2 above)
Comment (kbrandwijk) [reply packages/expo-cli/src/commands/publish/publishAsync.ts:172]: Behavior should be similar to how we ask about installing expo-dev-client (on a managed project only) when eas build kicks off a build with `developmentClient: true`. There's a detection there for managed/bare as well with behavior like you describe.

Comment (kbrandwijk) [reply packages/expo-cli/src/commands/publish/publishAsync.ts:172]: https://github.com/expo/eas-cli/blob/main/packages/eas-cli/src/build/utils/devClient.ts#L16

File: packages/expo-cli/src/commands/publish/publishAsync.ts (hunk 4 above)
Comment (kbrandwijk) [reply packages/expo-cli/src/commands/publish/publishAsync.ts:169]: Referring to `expo-dev-client` detecting in eas, that's simply done by finding `expo-dev-client/package.json` beneath the project root (so in node_modules): https://github.com/expo/eas-cli/blob/main/packages/eas-cli/src/build/utils/devClient.ts#L91-L99

File: packages/expo-cli/src/commands/publish/publishAsync.ts [hunk 5]
```diff
@@ -150,6 +154,25 @@ function assertValidReleaseChannel(releaseChannel?: string): void {
...
+
+  throw new Error(
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/publish/publishAsync.ts:171]:
```suggestion
  throw new CommandError(
```
Use ⌘ error so it logs pretty in the CLI

File: packages/expo-cli/src/commands/publish/publishAsync.ts [hunk 6]
```diff
@@ -150,6 +154,25 @@ function assertValidReleaseChannel(releaseChannel?: string): void {
...
+function assertExpoUpdatesInstalled(projectDir: string, sdkVersion?: string): void {
+  // before sdk 44, expo-update was included in with the expo module
+  if (sdkVersion && semver.lt(sdkVersion, '44.0.0')) {
+    return;
+  }
+
+  const packageJsonPath = resolveFrom.silent(projectDir, './package.json');
+  if (packageJsonPath) {
+    const expoPackageJson = JsonFile.read(packageJsonPath, { json5: true });
+    if (expoPackageJson.dependencies && (expoPackageJson.dependencies as any)['expo-updates']) {
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/publish/publishAsync.ts:166]:
```suggestion
    if ((expoPackageJson.dependencies as any)?.['expo-updates']) {
```
Or something of this nature

File: packages/expo-cli/src/commands/publish/publishAsync.ts (hunk 3 above)
Comment (EvanBacon) [inline packages/expo-cli/src/commands/publish/publishAsync.ts:159]: Probably want to use Versions from xdl since it accounts for `UNVERSIONED` and asserts when the semver is invalid in a controlled way.

**Comment by jkhales** [comment]
> i believe we should warn and prompt instead of hard fail here to account for one of the possible strange edge cases, so we don't block anybody.

Followed the confirmAsync pattern @kbrandwijk suggested.

**Review by brentvatne — APPROVED** [review]
